controllers/awscluster_controller_test.go

```diff
@@ -17,38 +17,1075 @@ limitations under the License.
+		mockCtrl = gomock.NewController(t)
+		recorder = record.NewFakeRecorder(10)
+		reconciler = AWSClusterReconciler{
+			Client:   testEnv.Client,
+			Recorder: recorder,
+		}
 	}
 
-	instance := &infrav1.AWSCluster{ObjectMeta: metav1.ObjectMeta{Name: "foo", Namespace: "default"}}
-	instance.Default()
+	teardown := func() {
+		mockCtrl.Finish()
+	}
+
+	t.Run("Should not create VPC in case of unmanaged VPC", func(t *testing.T) {
+		g := NewWithT(t)
+		mockCtrl = gomock.NewController(t)
+		ec2Mock := mock_ec2iface.NewMockEC2API(mockCtrl)
+		elbMock := mock_elbiface.NewMockELBAPI(mockCtrl)
+		expect := func(m *mock_ec2iface.MockEC2APIMockRecorder, e *mock_elbiface.MockELBAPIMockRecorder) {
+			mockedCreateVPCCalls(m)
+			mockedCreateSGCalls(m)
+			mockedCreateLBCalls(e)
+			mockedDescribeInstanceCall(m)
+		}
+		expect(ec2Mock.EXPECT(), elbMock.EXPECT())
+
+		capaNS := &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "capa-system",
+			},
+		}
+		g.Expect(testEnv.Create(ctx, capaNS)).To(Succeed())
+
+		controllerIdentity := createControllerIdentity(g)
+		ns, err := testEnv.CreateNamespace(ctx, fmt.Sprintf("integ-test-%s", util.RandomString(5)))
+		g.Expect(err).To(BeNil())
+		setup(t)
+		awsCluster := getAWSCluster("test", ns.Name)
+
+		g.Expect(testEnv.Create(ctx, &awsCluster)).To(Succeed())
+		defer teardown()
+		defer t.Cleanup(func() {
+			g.Expect(testEnv.Cleanup(ctx, &awsCluster, controllerIdentity, ns)).To(Succeed())
+		})
+
+		cs, err := getClusterScope(awsCluster)
+		g.Expect(err).To(BeNil())
+		networkSvc := network.NewService(cs)
+		networkSvc.EC2Client = ec2Mock
+		reconciler.networkServiceFactory = func(clusterScope scope.ClusterScope) services.NetworkInterface {
+			return networkSvc
+		}
+
+		ec2Svc := ec2Service.NewService(cs)
+		ec2Svc.EC2Client = ec2Mock
+		reconciler.ec2ServiceFactory = func(scope scope.EC2Scope) services.EC2MachineInterface {
+			return ec2Svc
+		}
+		testSecurityGroupRoles := []infrav1.SecurityGroupRole{
+			infrav1.SecurityGroupBastion,
+			infrav1.SecurityGroupAPIServerLB,
+			infrav1.SecurityGroupLB,
+			infrav1.SecurityGroupControlPlane,
+			infrav1.SecurityGroupNode,
+		}
+		sgSvc := securitygroup.NewService(cs, testSecurityGroupRoles)
+		sgSvc.EC2Client = ec2Mock
+
+		reconciler.securityGroupFactory = func(clusterScope scope.ClusterScope) services.SecurityGroupInterface {
+			return sgSvc
+		}
+		elbSvc := elbService.NewService(cs)
+		elbSvc.EC2Client = ec2Mock
+		elbSvc.ELBClient = elbMock
+
+		reconciler.elbServiceFactory = func(elbScope scope.ELBScope) services.ELBInterface {
+			return elbSvc
+		}
+		cs.SetSubnets([]infrav1.SubnetSpec{
+			{
+				ID:               "subnet-2",
+				AvailabilityZone: "us-east-1c",
+				IsPublic:         true,
+				CidrBlock:        "10.0.11.0/24",
+			},
+			{
+				ID:               "subnet-1",
+				AvailabilityZone: "us-east-1a",
+				CidrBlock:        "10.0.10.0/24",
+				IsPublic:         false,
+			},
+		})
+		_, err = reconciler.reconcileNormal(cs)
```

Comment:
Creating objects in envtest should trigger regular reconcile, so we should not need to make these calls explicitly.

If deletion ts is zero, it will go to reconcileNormal.

Reply:
if we do this, then how do we inject mocks in these calls for the services used. For eg. [here](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/37fd3c9fdb408e402eb921839dd196f745285c1a/controllers/awscluster_controller_test.go#L188) we are mocking EC2 service, but how do we inject that mock into the network service

Reply:
ACK, no need to try to find a workaround for this as there will be a refactoring after #3181.

Reply:
Will the refactoring resolve this issue? As still we won't be calling reconcile func and thus end up in same situation as explained above

Reply:
There are ways to solve this:
https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/8f04b857d07c8f34c3c0b0f18a6c9dba73b68875/exp/instancestate/awsinstancestate_controller_test.go#L106

With refactoring on the reconcilers, the above example could be cleaner. 
But feel free to do necessary refactoring in this PR itself.

Reply:
Ok looks like above workaround works good, I may have to refactor some tests a bit, so we can  let this PR open for some time